### PR TITLE
GRAU-31

### PR DESCRIPTION
**Pull request by DanMaguireData**

### CODE REVIEW

File: grau/utils/filter.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        Returns values to be passed to query.filter directly
+        """
+        if self.operator == StringOperators.CONTAINS.value:
+            return getattr(self.cls_, self.field).like(f"%{self.value}%")
+        if self.operator == StringOperators.STARTS_WITH.value:
+            return getattr(self.cls_, self.field).like(f"{self.value}%")
+        if self.operator == StringOperators.ENDS_WITH.value:
+            return getattr(self.cls_, self.field).like(f"%{self.value}")
+        if self.operator == StringOperators.LIKE.value:
+            return getattr(self.cls_, self.field)(self.value)
```
Comment (DANLENEHAN) [inline grau/utils/filter.py:121]: should this be ` getattr(self.cls_, self.field).like(self.value)`

File: grau/blueprints/stat/functions.py [hunk 2]
```diff
@@ -73,6 +73,10 @@ def get_stat(
     if max_timestamp is not None:
         stats = stats.filter(Stat.created_at <= max_timestamp)
 
+    value_is_null = parameters.get("value_is_null")
```
Comment (DANLENEHAN) [inline grau/blueprints/stat/functions.py:76]: Is this custom for this endpoint/function?

File: grau/utils/filter.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    )
+
+    def get_filter(self):
+        """
+        Returns values to be passed to query.filter directly
+        """
+        if self.operator == BooleanOperators.IS_TRUE.value:
+            return getattr(self.cls_, self.field) is True
+        if self.operator == BooleanOperators.IS_FALSE.value:
+            return getattr(self.cls_, self.field) is False
```
Comment (DANLENEHAN) [inline grau/utils/filter.py:174]: These is Bool filters don't work with the postgres dialect. You can print the actual sql the sqlachemy query produces with the following

from sqlalchemy.dialects import postgresql
print(query.statement.compile(dialect=postgresql.dialect()))

So for example 

query = db_session.query(User).filter(User.premium is True)

produces:

SELECT user_account.id, user_account.username, user_account.email, user_account.password, user_account.session_id, user_account.status, user_account.profile_link, user_account.premium, user_account.age, user_account.birthday, user_account.first_name, user_account.last_name, user_account.gender, user_account.phone_number, user_account.height_unit_pref, user_account.weight_unit_pref, user_account.date_format_pref, user_account.language, user_account.created_at, user_account.updated_at 
FROM user_account 
WHERE false

So you get no results

Comment (DANLENEHAN) [reply grau/utils/filter.py:174]: == True
or
== False
works:

Screenshot 2023-08-12 at 09 08 25

File: grau/grau_enums/filter_enums.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+
+class BaseOperators(Enum):
+    """
+    Base operators for filters.
+    """
+
+    EQUAL = "eq"
+    NOT_EQUAL = "ne"
+    IN = "in"
```
Comment (DANLENEHAN) [inline grau/grau_enums/filter_enums.py:11]: Could we also have "NOT_IN"

File: grau/utils/filter.py (hunk 1 above)
Comment (DANLENEHAN) [reply grau/utils/filter.py:121]: Although this .like with no %% seems to be just direct matching for me. E.g. the weight_unit_pref field .like("lbs") works but .like("lb") or .like("bs") so it seems like StringOperators.CONTAINS.value is really what we want for contains and like?

File: grau/utils/filter.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        """
+        Returns values to be passed to query.filter directly
+        """
+        if self.operator in [
+            DateTimeOperators.YEAR.value,
+            DateTimeOperators.MONTH.value,
+            DateTimeOperators.DAY.value,
+            DateTimeOperators.HOUR.value,
+            DateTimeOperators.MINUTE.value,
+            DateTimeOperators.SECOND.value,
```
Comment (DANLENEHAN) [inline grau/utils/filter.py:146]: So if we wanted to filter on all of these it would be 6 filter objects?

File: tests/utils/functions.py
Comment (DANLENEHAN) [inline tests/utils/functions.py:1]: Are alot of these functions not already in 'grau/utils.py'

**Review by DANLENEHAN — COMMENTED** [review]
Are we removing all the filtering on the get_stat endpoint and replacing it with just an id for now? Then I think in a subsequent ticket we can apply this new functionality with a new endpoint potentially to the stat case. In the same ticket we can probably also figure out how the docs will look for these.

File: grau/blueprints/stat/functions.py [hunk 6]
```diff
@@ -121,26 +78,47 @@ def update_stat(
...
+    """
+    QuerySchema.Config.context = {
+        "allowed_columns": {column.key for column in Stat.__table__.columns}
+    }
+    query_dict = SchemaValidator(QuerySchema).validate(query_dict)
+
+    query = Query(query=db_session.query(Stat), model=Stat)
+    query.apply_filters(query_dict.get("filters", {}))
+    query.apply_sort(query_dict.get("sort", {}))
+    return query.execute()
```
[the remark this answers is not in the document]
Comment (DanMaguireData) [reply grau/blueprints/stat/functions.py:124]: Just fixed this here, just want to know your opinion basically now it converts the list of objects into a dictionary of forms like:
{
  index: {dict},
  i.e. 0: {"name": "dan",....}
 }
 
 could do the id of the model as the key either if that is more intuitive

**Pull request merged**